### PR TITLE
Get GitHub diff using GitHub API

### PR DESCRIPTION
**Pull request by val-fed**

### CODE REVIEW

File: tools/reviewer/reviewer.proto [hunk 1]
```diff
@@ -44,45 +44,45 @@ message ReviewerConfig {
...
-    string id = 1;
-    string display_name = 2;
-    string project_url = 3;
-    // Short description
-    string description = 4;
-    // This is redundant since User lists projects, but we don't want to do the calculation
-    // in the front-end. We might add a test to enforce this, or push a better data structure to
-    // Firebase.
-    repeated string user_id = 5;
+  // A unique identifier for a user, e.g GitHub user
```
Comment (oferb) [inline tools/reviewer/reviewer.proto]: Interesting. Is the change only formatting? We have a formatter in CI, wonder why it didn't fail before, or now.

File: tools/reviewer/local_server/service/code_review.proto [hunk 2]
```diff
@@ -107,7 +107,7 @@ message Thread {
     CODE = 0;
     // Thread on commit (synced from other systems e.g GitHub)
     COMMIT = 1;
-    // Thread on repo (synced from other systems e.g GitHub)    
+    // Thread on repo (synced from other systems e.g GitHub)
```
Comment (oferb) [inline tools/reviewer/local_server/service/code_review.proto]: Similarly to below, wonder why formatter CI didn't fail.

File: tools/reviewer/job/sync/SyncTool.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ */
+public class SyncTool {
+  public static void main(String[] args) throws IOException {
+    Flags.parseCurrentPackage(args);
+
+    String repo;
+    int diffNumber;
+    String login;
+    String password;
+    try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
```
Comment (oferb) [inline tools/reviewer/job/sync/SyncTool.java]: Please make these inputs flags.

File: tools/reviewer/job/sync/SyncTool.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import com.google.startupos.common.flags.Flags;
+import com.google.startupos.tools.reviewer.localserver.service.Protos;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+/*
+ * Usage:
+ * bazel build //tools/reviewer/job/sync:sync_tool
```
Comment (oferb) [inline tools/reviewer/job/sync/SyncTool.java]: `bazel run` also builds, so no need for a separate `build` command.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+import org.kohsuke.github.GHPullRequestReviewComment;
+import org.kohsuke.github.GHRepository;
+import org.kohsuke.github.GitHub;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class RecipientGitHubDiff {
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: s/RecipientGitHubDiff/GitHubReader

We will have `GitHubReader`, `ReviewerReader` and `GitHubSync` that uses them both, and then syncs Reviewer and GitHub.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  private GHPullRequest pullRequest;
+  private List<GHPullRequestReviewComment> commentList;
+  private String repoName;
+
+  public RecipientGitHubDiff(String repoName, int diffNumber, String login, String password)
+      throws IOException {
+    this.repoName = repoName;
+    GitHub gitHub = GitHub.connectUsingPassword(login, password);
+    GHRepository repository = gitHub.getRepository(repoName);
+    pullRequest = repository.getPullRequest(diffNumber);
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: Let's do the actual reading from GitHub in a method, not in the constructor. That way, we could have a `GitHubReader` that can read multiple PRs.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    commentList = pullRequest.listReviewComments().asList();
+  }
+
+  public Protos.Diff getDiff() throws IOException {
+
+    Protos.Diff.Builder diff = Protos.Diff.newBuilder();
+
+    // diff.setId();
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: Can you check if there's a place to store `metadata` in a GitHub PR?
Initially, we will only support syncs for GitHub PRs that are initially created in Reviewer, so we will create them in GitHub and can put the diff number there, if there's a place we can put it.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    diff.setAuthor(Protos.Author.newBuilder().setEmail(pullRequest.getUser().getEmail()).build());
+
+    //    diff.addReviewer(Protos.Reviewer.newBuilder()
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: We won't sync reviewers initially. You can add a note about that.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    //        .setEmail(pullRequest.listReviews().asList().get(1).)
+    //        .setNeedsAttention()
+    //        .setApproved()
+    //        .build());
+    diff.setDescription(pullRequest.getBody());
+    //    diff.setBug();
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: We won't sync bug, status and workspace initially, you can also add a note for that.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    diff.setCreatedTimestamp(pullRequest.getCreatedAt().getTime());
+    diff.setModifiedTimestamp(pullRequest.getUpdatedAt().getTime());
+    diff.addAllCodeThread(getCodeThreads());
+    diff.addDiffThread(getDiffThread());
+
+    return diff.build();
+  }
+
+  private Protos.Thread getDiffThread() throws IOException {
+    List<GHIssueComment> issueComments = pullRequest.getComments();
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: s/issueComments/diffComments

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+  private List<GHPullRequestReviewComment> commentList;
+  private String repoName;
+
+  public RecipientGitHubDiff(String repoName, int diffNumber, String login, String password)
+      throws IOException {
+    this.repoName = repoName;
+    GitHub gitHub = GitHub.connectUsingPassword(login, password);
+    GHRepository repository = gitHub.getRepository(repoName);
+    pullRequest = repository.getPullRequest(diffNumber);
+    commentList = pullRequest.listReviewComments().asList();
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: s/commentList/codeComments

File: tools/reviewer/job/sync/RecipientGitHubDiff.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+            e.printStackTrace();
+          }
+        });
+    return thread.build();
+  }
+
+  private List<Protos.Thread> getCodeThreads() {
+
+    List<Protos.Thread> result = new ArrayList<>();
+    List<String> comentedFiles =
```
Comment (oferb) [inline tools/reviewer/job/sync/RecipientGitHubDiff.java]: s/comentedFiles/commentedFiles

File: tools/reviewer/reviewer.proto (hunk 1 above)
Comment (val-fed) [reply tools/reviewer/reviewer.proto]: Yes, `formatter` made these changes.

File: tools/reviewer/job/sync/SyncTool.java (hunk 3 above)
Comment (val-fed) [reply tools/reviewer/job/sync/SyncTool.java]: Done.

File: tools/reviewer/job/sync/SyncTool.java (hunk 4 above)
Comment (val-fed) [reply tools/reviewer/job/sync/SyncTool.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 5 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 6 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 8 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 9 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 10 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 11 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 12 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Done.

File: tools/reviewer/job/sync/RecipientGitHubDiff.java (hunk 7 above)
Comment (val-fed) [reply tools/reviewer/job/sync/RecipientGitHubDiff.java]: Looks like it is impossible.
As a variant, we can add a label with `diff number`.

**Pull request closed without merging**